### ISSUE TITLE
Miscompilation on M1 macs with macOS Ventura and Apple Clang?

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#5385]
Anton Betten observed the following trying to start a self-compiled GAP 4.12.2:
```
% gap
┌───────┐   GAP 4.12.2 of 2022-12-18
│  GAP  │   https://www.gap-system.org/
└───────┘   Architecture: aarch64-apple-darwin22-default64-kv8
Configuration:  gmp 6.2.1, GASMAN
Loading the library and packages ...
Syntax warning: Unbound global variable in .../gap-4.12.2/lib/primality.gi:515
 Np:=(N-1)/p;
 ^^
Syntax warning: Unbound global variable in .../gap-4.12.2/lib/primality.gi:515
 Np:=(N-1)/p;
      ^
...
Syntax warning: Unbound global variable in .../gap-4.12.2/lib/primality.gi:527
  return [true,a];
               ^
Segmentation fault
%
```
He reported the following configuration details about this computer, and also that the issue goes away when he compiles with GCC 12 (installed via Homebrew) instead of using cc.
```
% sw_vers
ProductName:    macOS
ProductVersion: 13.0
BuildVersion:   22A380

% cc --version
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

% xcode-select -p
/Applications/Xcode.app/Contents/Developer

% which cc
/usr/bin/cc
```

@hulpke added:

> I have seen this problem on three separate machines:
> 
> An M1 Pro 14” MacBook Pro
> 
> An M1 Max Mac Studio
> 
> An M2 MacBook Air
> 
> It comes in conjunction with frequent segfaults on calculations. The errors started to arise in Fall, so it could be a Ventura/compiler version issue. It happens also if I reduce optimization level to 1.
> 
> Compiling with (real) gcc makes the errors go away.

Since I build and use GAP daily on macOS 12.6.3, with Apple clang version 14.0.0 (clang-1400.0.29.202) (so the same as Anton reported), the one common factor seems to be that both he and @hulpke use Ventura.

I'll try and see if I can get access to a machine running Ventura, to test if I can reproduce.

**Comment by hulpke** [comment]
Dear @fingolfin 

If you want I can give you an account on my machine. (M1Max Mac Studio)

Best,

  Alexander

**Comment by ChrisJefferson** [comment]
@fingolfin is more of a mac expert than me, but I'd be happy to take a look if someone gave me access to a machine where the crash occurs.

**Comment by ChrisJefferson** [comment]
Also, taking a wild guess, Mac has in the past changed how setjmp/longjmp is handled, so that would be the first place I'd look, in gasman -- although I might expect this code to be in the library, not change between compilers.

**Comment by fingolfin** [comment]
@hulpke that would be helpful, yes

**Comment by dimpase** [comment]
@sagemath gets Apple-specific reports of this sort too, but so far it always boils down to a toolchain setup, in particular mixing arm64 (aka M1/M2) and x86_64 emulation.

**Comment by fingolfin** [comment]
Yeah I am suspicious about this too, which is why it'd be great to get SSH access to a machine which reproduces this, for analysis. It seems strange that the identical compiler would produce a binary that runs on macOS 12 but fails on 13 (not impossible, mind you, but it doesn't sound like simple miscompilation then)

**Comment by dimpase** [comment]
> binary that runs on macOS 12 but fails on 13

all that `MACOSX_DEPLOYMENT_TARGET` voodoo

**Comment by dimpase** [comment]
given a rather opaque way Apple distributes their upgrades, your "stock" might differ from someone's else on a very similar system. And sometimes what you get from Apple's "command line tools" differs from what you get in XCode.

I'd suggest to specify the exact version. 

By the way, Homebrew provides its own clang, usually a newer one. I recall building GAP 4.11 with it,it all worked and tested well.

**Comment by hulpke** [comment]
This is what I have observed (on several machines, but only after upgrading to Ventura), when uding GAP compiled with Apples standard tools. All of this happens sometimes out of the blue and might go away if I just restart the same Job again. (This makes me think that it has to do with memory management or where GAP exactly is located in memory.):

- An error, as Anton observed when starting up and loading primality.gi. If I restart a few times it eventually goes away.

- A sudden segfault in a larger calculation, typically wityhin an hour or two, but not necessarily in the first 10 minutes. The location when changes. FWIW, it involved the Dixon-Schneider and thus linear algebra over larger finite fields.

**Comment by fingolfin** [comment]
So far I was not able to reproduce the crash on your machines.

Do you have any particular code snippet that runs for a long time that "often" triggers it? I've now compiled both GAP master and GAP 4.12.2 on your machine, and so far they work great: I've been repeatedly starting and exiting them, running the testinstall testsuite etc., each dozens of times.

I also tried to simulate some "pressure" by running 10 instances of the testsuite in parallel: `for i in 1 2 3 4 5 6 7 8 9 0 ; do   ./gap tst/testinstall.g & done`. Still not  a single crash :-(.


Maybe there is still a difference in what we are doing, gotta figure out what that might be...
- For the GAP 4.12.2 build, it compiled its own GMP -- did you or Anton instruct it to use a specific GMP? Or is it  using the bundled GMP?
- Did you build any of the packages?
- are perhaps packages being loaded other than those bundled with GAP?
- 
Perhaps you could email me your config.log files, that might reveal something.

**Comment by fingolfin** [comment]
@hulpke also, if you you send me the `gap` executable that exhibits the crashes, that would be super helpful

**Comment by hulpke** [comment]
I now have been unable to reproduce it either. No idea whether the compiler changed. I will try again in a little while.

**Comment by fingolfin** [comment]
We can't reproduce this yet many people are building and using GAP on M1 and M2 macs by now. Closing this -- if anyone sees this again, just open a new issue
